agent/app/agent.go

```diff
@@ -0,0 +1,441 @@
+	if err != nil {
+		log.Criticalf("Error creating state manager: %v", err)
+		return exitcodes.ExitTerminal
+	}
+
+	// Register the container instance
+	err = agent.registerContainerInstance(taskEngine, stateManager, client)
+	if err != nil {
+		if isTranisent(err) {
+			return exitcodes.ExitError
+		}
+		return exitcodes.ExitTerminal
+	}
+
+	// Begin listening to the docker daemon and saving changes
+	taskEngine.SetSaver(stateManager)
+	imageManager.SetSaver(stateManager)
+	taskEngine.MustInit(agent.ctx)
+
+	// Start back ground routines, including the telemetry session
+	deregisterInstanceEventStream := eventstream.NewEventStream(
+		deregisterContainerInstanceEventStreamName, agent.ctx)
+	deregisterInstanceEventStream.StartListening()
+	taskHandler := eventhandler.NewTaskHandler()
+	agent.startAsyncRoutines(containerChangeEventStream, credentialsManager, imageManager,
+		taskEngine, stateManager, deregisterInstanceEventStream, client, taskHandler)
+
+	// Start the acs session, which should block doStart
+	return agent.startACSSession(credentialsManager, taskEngine, stateManager,
+		deregisterInstanceEventStream, client, taskHandler)
+}
+
+// newTaskEngine creates a new docker task engine object. It tries to load the
+// local state if needed, else initializes a new one
+func (agent *ecsAgent) newTaskEngine(containerChangeEventStream *eventstream.EventStream,
+	credentialsManager credentials.Manager,
+	state dockerstate.TaskEngineState,
+	imageManager engine.ImageManager) (engine.TaskEngine, string, error) {
+
+	containerChangeEventStream.StartListening()
+
+	if !agent.cfg.Checkpoint {
+		log.Info("Checkpointing not enabled; a new container instance will be created each time the agent is run")
+		return engine.NewTaskEngine(agent.cfg, agent.dockerClient,
+			credentialsManager, containerChangeEventStream, imageManager, state), "", nil
+	}
+
+	// We try to set these values by loading the existing state file first
+	var previousCluster, previousEC2InstanceID, previousContainerInstanceArn string
+	previousTaskEngine := engine.NewTaskEngine(agent.cfg, agent.dockerClient,
+		credentialsManager, containerChangeEventStream, imageManager, state)
+
+	// previousState is used to verify that our current runtime configuration is
+	// compatible with our past configuration as reflected by our state-file
+	previousState, err := agent.newStateManager(previousTaskEngine, &previousCluster,
+		&previousContainerInstanceArn, &previousEC2InstanceID)
+	if err != nil {
+		log.Criticalf("Error creating state manager: %v", err)
+		return nil, "", err
+	}
+
+	err = previousState.Load()
+	if err != nil {
+		log.Criticalf("Error loading previously saved state: %v", err)
+		return nil, "", err
+	}
+
+	if previousCluster != "" {
+		if err := agent.setClusterInConfig(previousCluster); err != nil {
+			return nil, "", err
+		}
+	}
+
+	currentEC2InstanceID := agent.getEC2InstanceID()
+	if previousEC2InstanceID != "" && previousEC2InstanceID != currentEC2InstanceID {
```

Comment:
Yes, I believe that the current behavior is a bug. If the EC2 instance ID is different, it shouldn't matter that the cluster name is mismatched because we'd want to reset the task engine anyway.